staging/src/k8s.io/apiserver/pkg/endpoints/handlers/workflow.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handlers
+
+import (
+	"net/http"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// buildWorkflowID builds a reasonable default for the workflowID, based
+// on user-agent of the incoming request, sub-resource, verb and
+// content-type. Always returns a non-empty string.
+// TODO(apelisse): We need a way to clean-up obsolete workflow and/or
+// override defaulted workflow with specific workflow.
+// TODO(apelisse): We also probably need to understand how changing this
+// heuristic would impact object that use former values.
+func buildDefaultWorkflowID(subResource, verb, contentType, userAgent string) string {
+	// This is the worst-case scenario: we don't have anything useful
+	workflowID := "default"
+
+	if userAgent != "" {
+		workflowID = strings.SplitN(userAgent, "/", 2)[0]
+	}
+
+	if subResource != "" {
+		workflowID = workflowID + "-" + subResource
+	}
+
+	if !(verb == http.MethodPost || (verb == http.MethodPatch && types.PatchType(contentType) == types.ApplyPatchType)) {
```

Comment:
suggest making this a switch statement with a number of empty cases that catch the various conditions--humans weren't built to read that much boolean algebra in one line...

Reply:
Yeah, I'll update, not sure it's going to be much clearer.